Change timeline chart and add start, end time for context chart

karelhala opened the pull request:
In Timeline chart the date was wrongly formatted, used D instead of d for days. Also data were not correctly shown in graph, domain for timeline scale did't started at startTime of graph, but on fisrt timestamp of data, same applies for end of graph.

MiQ graphs already contains html partial so there is now option to pass html instead of data object.

Add new start and end of brush to context chart, this way it's possible to change the brush with filtering inside different chart.

karelhala committed:
Change timeline chart and add start, end time for context chart
In Timeline chart the date was wrongly formatted, used D instead of d for days. Also data were not correctly shown in graph, domain for timeline scale did't started at startTime of graph, but on fisrt timestamp of data, same applies for end of graph.

Add new start and end of brush to context chart, this way it's possible to change the brush with filtering inside different chart.

mtho11 commented on src/chart/timeline-directive.ts:
```diff
@@ -303,12 +318,13 @@ namespace Charts {
               return 3;
-            }) .on('mouseover', (d, i) => {
-              tip.show(d, i);
+            }) .on('mouseover', function(d, i) {
```
No need for "function" not sure why you changed from "=>" syntax

karelhala replied:
I changed it to function because I needed to address currently changed circle. Based on d3 documentation they call these function with `this` set to current circle. But good point, I've changed it so it calles `positionTip` function only and selecting of current circle happens inside `positionTip` function.

mtho11 commented on src/chart/timeline-directive.ts:
```diff
@@ -147,8 +150,8 @@ namespace Charts {
         // data specific vars
-        let startTimestamp: number = +attrs.startTimestamp,
-          endTimestamp: number = +attrs.endTimestamp,
+        let startTimestamp: number = +scope.startTimestamp,
```
There should be no 'scope', with angular 1.5 there are only a very few cases where its still needed. This is in preparation for Angular 2.0.

karelhala replied:
Good point, I'll revert it back to attrs.

mtho11 commented on src/chart/context-chart-directive.ts:
```diff
@@ -64,9 +66,9 @@ namespace Charts {
 
-          width = (<any>parentNode).clientWidth;
+          //let's use 92.5% of parents width
+          width = (<any>parentNode).clientWidth * 0.925;
```
There are times when we want the charts to exactly line up with other charts (for example: a context chart to align with a event timeline and metrics chart). Having pixel alignment with a margin in pixels better facilitates this instead of %.

karelhala replied:
I've added this here because I was not aware of transformation inside `svg`. But I'll move it to `tranform` function which is called couple of lines bellow.
The reason why I did this was because the graph did not fit parent element and part of this graph was missing.

mtho11 commented on src/chart/context-chart-directive.ts:
```diff
@@ -226,6 +227,14 @@ namespace Charts {
           }
         }
 
+        function redrawBrush(startTimestamp, endTimestamp) {
```
Timestamp have a type of TimeInMillis for documentation and type safety.

mtho11 commented on src/chart/context-chart-directive.ts:
```diff
@@ -226,6 +227,14 @@ namespace Charts {
+          if (brush) {
+            brush.extent([new Date(startTimestamp), new Date(endTimestamp)]);
+            brush(d3.select('hk-context-chart .brush').transition());
+            brush.event(d3.select('hk-context-chart .brush').transition());
```
There is a space in 'hk-context-chart .brush'

karelhala committed:
Change calling of positionTip, change scaling of context chart
when tip is positioned in timeline directive there is no need, to use function and call another function, instead call just one function and rely on d3 documentation, that  is currently edited element.

When using context chart we align it by 90 pixels to left, so we have to scale the graph to 90% (on x axis, y axis is kept at 100%) so it will fit the data and inside parent's area.

karelhala committed:
Change scope to attr in addressing start and endtimestamp for timeline directive
